### PR TITLE
deis doctor includes lots of k8s data

### PR DESCRIPTION
**Pull request by jackfrancis**
This PR delivers the following feature additions:

- Deis doctor handler introspects the configured "deis" kubernetes namespace and generates a JSON-body POST request to the configured deis doctor API endpoint
- Kuberenetes data gathered and reported by deis doctor includes:
  - nodes
  - pods (namespaced to "deis")
  - daemonSets (namespaced to "deis")
  - deployments (namespaced to "deis")
  - events (namespaced to "deis")
  - replicaSets (namespaced to "deis")
  - replicationControllers (namespaced to "deis")
  - services (namespaced to "deis")
- Kubernetes data is gathered *au totale* by way of a swagger model property that eagerly JSON-encodes the entire source kubernetes resource data inside a `data` property. This makes for a predictably heavy-weight data payload, and in the future it may make sense to include some filtering. That, and any additional explicit properties data modeling is TODO.

Included in the implementation to support these features are the following under-the-hood items worth emphasizing:

- A new `k8s` sub-package was created to accommodate all new business logic responsible for gathering and preparing kubernetes cluster data; pre-existing kubernetes-specific implementations that were formerly in the `data` package were moved to `k8s`, as appropriate (e.g., the `data.getRCItems()` helper function becomes the exportable `k8s.GetReplicationControllers()`.
- To support the new kubernetes-specific interfaces needed, the handy `kubeapp` dependency was updated, and in fact `github.com/deis` absorbed what was formerly @arschles's personal creation.
- `glide.yaml` and `glide.lock` updated to reflect updated `kubeapp` dependencies
- Zero-value k8s data mocks added to support very basic contract-abiding test interfaces, but actual mock data is TODO in a follow-up PR.
- Resource-specific interfaces added as light wrappers to aid the passing around of data and data specifications. In particular, the `k8s.ResourceInterface` is meant, with the help of its constructor, to conveniently provide a re-usable kubernetes-namespace-specific, per-resource data retrieval singleton.
- New, swagger-generated client stubs to support deis doctor k8s data.

### CODE REVIEW

**Comment by arschles** [comment]
moving out of the 2.2 milestone

File: data/data.go [hunk 1]
```diff
@@ -31,7 +32,7 @@ func GetCluster(
 	}
 	err = AddUpdateData(&cluster, v, secretGetterCreator)
 	if err != nil {
-		return models.Cluster{}, err
+		log.Println("unable to decorate cluster data with available updates data")
```
Comment (jackfrancis) [inline data/data.go]: When we're unable to add definitive "available component update" data we'll log the event, but we don't want to short-circuit the entire version data-gathering process.

**Comment by jackfrancis** [comment]
Above-noted future TODO item for k8s mock data filed here:

- https://github.com/deis/workflow-manager/issues/85

**Comment by jackfrancis** [comment]
Above-noted future TODO for deis doctor data modeling filed here:

- https://github.com/deis/workflow-manager/issues/86

File: data/installed_data_test.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package data
```
Comment (jackfrancis) [inline data/installed_data_test.go]: TODO: delete this placeholder file (or actually add tests).

File: data/available_versions_test.go [hunk 3]
```diff
@@ -13,6 +14,69 @@ import (
...
+	        "name": "bypass me",
+	        "description": "bypass me"
+	      },
+	      "version": {
+	        "version": "v2-bypass"
+	      }
+	    }
+	  ]
+	}]`))
+	_ = json.Unmarshal(data, &componentVersions)
```
Comment (arschles) [inline data/available_versions_test.go]: you should either return this error or just construct the `[]models.ComponentVersion` directly and return it

File: data/data.go [hunk 4]
```diff
@@ -21,17 +21,16 @@ func GetCluster(
...
 	// Populate cluster object with installed components
 	cluster, err := GetInstalled(c)
 	if err != nil {
 		return models.Cluster{}, err
 	}
-	err = AddUpdateData(&cluster, v, secretGetterCreator)
+	err = AddUpdateData(&cluster, v)
 	if err != nil {
-		return models.Cluster{}, err
+		log.Println("unable to decorate cluster data with available updates data")
```
Comment (arschles) [inline data/data.go]: 2 things:

1. why not return the error anymore?
2. this log message should include the error in it

File: data/data.go [hunk 5]
```diff
@@ -139,35 +136,80 @@ func NewestSemVer(v1 string, v2 string) (string, error) {
...
+	k k8s.RunningK8sData, // k8s data
 	i ClusterID,
 	v AvailableComponentVersion,
-	secretGetterCreator KubeSecretGetterCreator,
 ) (models.DoctorInfo, error) {
-	cluster, err := GetCluster(c, i, v, secretGetterCreator)
+	cluster, err := GetCluster(c, i, v)
 	if err != nil {
-		return models.DoctorInfo{}, err
+		log.Println("unable to get workflow cluster data")
```
Comment (arschles) [inline data/data.go]: 2 things:

1. why not return the error anymore?
2. this log message should include the error in it

File: data/data.go [hunk 6]
```diff
@@ -139,35 +136,80 @@ func NewestSemVer(v1 string, v2 string) (string, error) {
...
 	}
+	nodes := getK8sNodes(k)
+	var namespaces []*models.Namespace
+	namespaces = append(namespaces, getK8sDeisNamespace(k))
```
Comment (arschles) [inline data/data.go]: replace the above two lines with: `namespaces := []*models.Namespace{getK8sDeisNamespace(k)}`

File: data/data.go [hunk 7]
```diff
@@ -139,35 +136,80 @@ func NewestSemVer(v1 string, v2 string) (string, error) {
...
 	}
 	return doctor, nil
 }
 
+// getK8sDeisNamespace is a helper function that returns data
+// from the "deis" K8s namespace for RESTful consumption
+func getK8sDeisNamespace(k k8s.RunningK8sData) *models.Namespace {
+	pods, err := k8s.GetPodsModels(k)
+	if err != nil {
+		log.Println("unable to get K8s pods data")
```
Comment (arschles) [inline data/data.go]: why not return the error here and below? do you not care if the get operations fail?

File: data/data.go (hunk 4 above)
Comment (jackfrancis) [inline data/data.go]: 1. I don't want the failure of determining whether or not a newer component versions are available (what `AddUpdateData` does) to hijack the entire cluster data gathering process. I would prefer to simply return the cluster data without any update data decorations, from which the user would infer that available update data is undetermined.
2. will fix

File: data/data.go (hunk 5 above)
Comment (jackfrancis) [inline data/data.go]: 1. this is a mistake, will fix

File: handlers/handlers.go [hunk 8]
```diff
@@ -68,14 +65,14 @@ func ComponentsHandler(
...
+	c data.InstalledData, // deis cluster
+	k k8s.RunningK8sData, // k8s cluster
 	i data.ClusterID,
 	v data.AvailableComponentVersion,
-	secretGetterCreator data.KubeSecretGetterCreator,
 	apiClient *apiclient.WorkflowManager,
 ) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		doctor, err := data.GetDoctorInfo(c, i, v, secretGetterCreator)
+		doctor, err := data.GetDoctorInfo(c, k, i, v)
```
Comment (arschles) [inline handlers/handlers.go]: can we get some _slightly_ more descriptive variables here?

File: data/data.go (hunk 7 above)
Comment (jackfrancis) [inline data/data.go]: It's correct that we want to "log and continue" if errors occur during this data-gathering process. We want to get as much data as we can.

However, we should log the error, will fix.

File: jobs/jobs.go [hunk 9]
```diff
@@ -66,14 +63,12 @@ type getLatestVersionData struct {
 	installedData         data.InstalledData
 	clusterID             data.ClusterID
 	availableComponentVsn data.AvailableComponentVersion
-	sgc                   data.KubeSecretGetterCreator
+	k8sResources     k8s.ResourceInterfaceNamespaced
```
Comment (arschles) [inline jobs/jobs.go]: looks like this needs `go fmt`ing - the type should vertically align with the others in the struct

File: k8s/k8s.go [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+	podLister        pod.Lister
+	rcLister         rc.Lister
+	replicaSetLister replicaset.Lister
+	serviceLister    service.Lister
+}
+
+// NewRunningK8sData returns a new runningK8sData using rcl as the rc.Lister implementation
+func NewRunningK8sData(r *ResourceInterfaceNamespaced) RunningK8sData {
+	return &runningK8sData{daemonSetLister: r.DaemonSets(), deploymentLister: r.Deployments(), eventLister: r.Events(),
+		nodeLister: r.Nodes(), podLister: r.Pods(), rcLister: r.ReplicationControllers(), replicaSetLister: r.ReplicaSets(), serviceLister: r.Services()}
```
Comment (arschles) [inline k8s/k8s.go]: can you list each struct field on a new line

File: k8s/k8s.go [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+// DaemonSets method for runningK8sData
+func (rkd *runningK8sData) DaemonSets() ([]*models.K8sResource, error) {
+	ret := []*models.K8sResource{}
```
Comment (arschles) [inline k8s/k8s.go]: `ret` should be pre-allocated: `ret := make([]*models.K8sResource, len(ds))`

Comment (arschles) [inline k8s/k8s.go]: same with below allocations

File: k8s/k8s.go [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+		ret = append(ret, service)
+	}
+	return ret, nil
+}
+
+// GetNodesModels gets k8s node model data for RESTful consumption
+func GetNodesModels(k RunningK8sData) ([]*models.K8sResource, error) {
+	nodes, err := k.Nodes()
+	if err != nil {
+		return []*models.K8sResource{}, err
```
Comment (arschles) [inline k8s/k8s.go]: this should `return nil, nil`, to avoid unnecessarily allocating an empty slice

Comment (arschles) [inline k8s/k8s.go]: same with below `return`s in other funcs

File: data/available_versions_test.go [hunk 13]
```diff
@@ -20,7 +20,10 @@ type testAvailableVersions struct{}
 func (a testAvailableVersions) Refresh(cluster models.Cluster) ([]models.ComponentVersion, error) {
 	data := getMockComponentVersions()
 	var componentVersions []models.ComponentVersion
-	_ = json.Unmarshal(data, &componentVersions)
+	err := json.Unmarshal(data, &componentVersions)
+	if err != nil {
```
Comment (arschles) [inline data/available_versions_test.go]: combine these lines. `if err := json.Unmarshal(...); err != nil { ...`

Comment (arschles) [inline data/available_versions_test.go]: same with the below combinations

File: data/data.go [hunk 14]
```diff
@@ -30,7 +30,7 @@ func GetCluster(
 	}
 	err = AddUpdateData(&cluster, v)
 	if err != nil {
-		log.Println("unable to decorate cluster data with available updates data")
+		log.Printf("unable to decorate cluster data with available updates data: %#v\n", err)
```
Comment (arschles) [inline data/data.go]: remove the trailing `\n`

Comment (arschles) [inline data/data.go]: same with the below removals

File: k8s/k8s.go [hunk 15]
```diff
@@ -119,11 +127,11 @@ func (rkd *runningK8sData) Nodes() ([]*models.K8sResource, error) {
...
 // Pods method for runningK8sData
 func (rkd *runningK8sData) Pods() ([]*models.K8sResource, error) {
-	ret := []*models.K8sResource{}
 	pods, err := getPods(rkd.podLister)
 	if err != nil {
-		return ret, err
+		return nil, err
 	}
+	ret := make([]*models.K8sResource, len(pods))
 	for _, p := range pods {
```
Comment (arschles) [inline k8s/k8s.go]: use the index in this `range` to insert into `ret`:

```go
for i, p := range pods {
  // ...
  ret[i] = pod
}
```

Comment (arschles) [inline k8s/k8s.go]: same with the above and below `range`s

File: data/data.go [hunk 16]
```diff
@@ -28,9 +28,8 @@ func GetCluster(
 	if err != nil {
 		return models.Cluster{}, err
 	}
-	err = AddUpdateData(&cluster, v)
-	if err != nil {
-		log.Printf("unable to decorate cluster data with available updates data: %#v\n", err)
+	if err = AddUpdateData(&cluster, v); err != nil {
```
Comment (arschles) [inline data/data.go]: `if err := AddUpdateData...`. note the `:` in the `if`

**Comment by mboersma** [comment]
I tested this by putting it into the workflow-dev chart. Resulted in a huge JSON blob full of the above information--good stuff!

**Comment by jackfrancis** [comment]
closes #79

**Pull request merged**